### PR TITLE
Feature/ts 111 - Dynamicall rendering holiday booking modal

### PR DESCRIPTION
**Pull request by PaulTummonUnosquare**
Feature/ts 111 - Dynamicall rendering holiday booking modal

image

image

image

image

image

image

### CODE REVIEW

File: src/components/BookingModal/BookingModalForm/index.js [hunk 1]
```diff
@@ -24,29 +25,32 @@ const BookingModalForm = props => {
...
+  const { eventStatus: {eventStatusId} } = booking;
+  const isEventCancelled = eventStatusId === holidayStatus.CANCELLED;
+  const rejectionMessages = booking.messages && !isEventCancelled;
   const { eventTypeId } = formData;
   const buttonTextValue =
   `${eventTypeId !== eventTypes.ANNUAL_LEAVE ? '' : bookingDuration === 0.5 ? 'Half' : bookingDuration}
   ${eventTypeId !== eventTypes.ANNUAL_LEAVE ? 'WFH' : bookingDuration > 1 ? 'Days' : 'Day'}`;
 
   const createCtas = () => {
+    const isButtonDisabled = isEventCancelled ? isEventCancelled : !formIsValid;
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/index.js]: Can this be `const isButtonDisabled = isEventCancelled || !formIsValid`?

File: src/components/common/Form/index.js [hunk 2]
```diff
@@ -69,9 +69,11 @@ export class Form extends Component {
 
   render() {
     const childWithProp = React.Children.map(this.props.children, child => {
-      return React.cloneElement(child, {
-        validateForm: this.handleCheckInputValid,
-      });
+      if (child.type !== 'div') {
```
Comment (bgribz) [inline src/components/common/Form/index.js]: This seems like a bit of a code smell - why is this necessary?

File: src/components/BookingModal/BookingModalForm/index.js [hunk 3]
```diff
@@ -91,44 +95,48 @@ const BookingModalForm = props => {
     <Fragment>
       {composeErrorMessage()}
       <Form formData={formData} formStatus={formStatus} actions={createCtas()}>
-        <Input
+        {!rejectionMessages ? <Input
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/index.js]: All of this conditional rendering could be handled better - it seems like this form's entire contents depend on `rejectionMessages` being false, so maybe wrap it in a single ternary.

**Review by bgribz — CHANGES_REQUESTED** [review]

File: src/components/common/Form/index.js (hunk 2 above)
Comment (PaulTummonUnosquare) [reply src/components/common/Form/index.js]: as there are default empty divs returned within the form to dynamically build the elements as react was throwing an error that the validation doesn't apply to empty divs.

Comment (William-Unosquare) [reply src/components/common/Form/index.js]: I got a feeling the solution then is to dig into why the validation doesn't like these rules and fix it from there. Gonna take a closer look at this PR now.

File: src/components/BookingModal/BookingModalForm/index.js (hunk 1 above)
Comment (William-Unosquare) [inline src/components/BookingModal/BookingModalForm/index.js]: In the other refactor branch, TS-109, there is now a function called `isValidBooking` which drives the logic for this, as there are quite a few different conditions where the button should be disabled. I recommend merging that branch into this one and using it.

File: src/components/BookingModal/BookingModalForm/index.js [hunk 4]
```diff
@@ -24,29 +25,32 @@ const BookingModalForm = props => {
...
     if (isEventBeingUpdated) {
       return [
         {
-          label: `Update to ${
+          label: isEventCancelled ? 'Cancelled' : rejectionMessages ? 'Submit' : `Update to ${
             bookingDuration === 0.5 ? 'Half' : bookingDuration
           } ${bookingDuration > 1 ? 'Days' : 'Day'}`,
           event: updateEvent,
-          disabled: !formIsValid,
+          disabled: isButtonDisabled,
```
Comment (William-Unosquare) [inline src/components/BookingModal/BookingModalForm/index.js]: This will likely cause conflicts with the refactor branch because we're now driving this using the `isValidBooking` function in the container. Again, I recommend a merge.

**Review by William-Unosquare — CHANGES_REQUESTED** [review]

File: src/components/BookingModal/BookingModalForm/index.js (hunk 1 above)
Comment (bgribz) [reply src/components/BookingModal/BookingModalForm/index.js]: Emmm maybe wait until that other branch has made it into develop and merge from there just to keep each ticket's changes separate.

File: src/components/common/Form/index.js (hunk 2 above)
Comment (bgribz) [reply src/components/common/Form/index.js]: Can you not return `null` instead of empty divs?

File: src/components/BookingModal/BookingModalForm/index.js (hunk 1 above)
Comment (William-Unosquare) [reply src/components/BookingModal/BookingModalForm/index.js]: ^

File: src/components/BookingModal/BookingModalForm/index.js (hunk 3 above)
Comment (PaulTummonUnosquare) [reply src/components/BookingModal/BookingModalForm/index.js]: I don't think they can be blocked together with the logic as only some of the inputs depend independently upon only rejectionMessages. There are some others throughout the forms composition that depend upon cancelled event too and they don't render within a single block.

Comment (William-Unosquare) [reply src/components/BookingModal/BookingModalForm/index.js]: I think a separation of the main form & the rejection input will help. Put each in their own component and pick which to render 1 level up

File: src/components/BookingModal/AlertMessage/styled.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  position: relative;
+  top: 10px;
+  margin-right: 8px;
+}
+p {
+  margin: 0 0 5px 0;
+  :last-of-type {
+    margin-left: 26px;
+  }
+}
```
Comment (bgribz) [inline src/components/BookingModal/AlertMessage/styled.js]: I wonder if we're missing a linting rule for this, but can you tab this out?

File: src/components/BookingModal/BookingModalForm/container.js [hunk 6]
```diff
@@ -28,48 +31,69 @@ const Container = Wrapped =>
...
-          updateMessage: '',
+      this.setState(
+        {
+          formData: {
+            start: start,
+            end: end,
+            eventTypeId: this.state.workingFromHomeBooking
+              ? eventTypes.WFH
+              : eventType.eventTypeId,
+            isHalfday: halfDay || false,
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/container.js]: Can we get rid of the `|| false` here?

File: src/components/BookingModal/BookingModalForm/container.js [hunk 7]
```diff
@@ -28,48 +31,69 @@ const Container = Wrapped =>
...
+          },
         },
-      });
+        () => {
+          onFormUpdate(this.state.formData);
+        }
+      );
     };
 
+    WFHBooking() {
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/container.js]: Since this is returning a boolean it should be called something like `isWFHBooking()`

File: src/components/BookingModal/BookingModalForm/container.js [hunk 8]
```diff
@@ -28,48 +31,69 @@ const Container = Wrapped =>
...
+      const {
+        workingFromHomeBooking,
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/container.js]: Should probably keep our naming of "Working from home" consistent across the app - one of either `wfh` or `workingFromHome`.

File: src/components/common/Form/index.js [hunk 9]
```diff
@@ -69,9 +69,11 @@ export class Form extends Component {
 
   render() {
     const childWithProp = React.Children.map(this.props.children, child => {
-      return React.cloneElement(child, {
-        validateForm: this.handleCheckInputValid,
-      });
+      if (child) {
```
Comment (bgribz) [inline src/components/common/Form/index.js:72]: Is there an error thrown when this is removed?

**Review by bgribz — CHANGES_REQUESTED** [review]
A few things, plus we need to sort those line endings.

**Comment by mcromie92** [comment]
@bgribz @PaulTummonUnosquare @William-Unosquare fixed some merge issues so can I get this reviewed again please :)

File: src/components/BookingModal/BookingModalForm/index.js [hunk 10]
```diff
@@ -44,19 +49,17 @@ const BookingModalForm = props => {
     if (isEventBeingUpdated) {
       return [
         {
-          label: booking.messages
-            ? 'Submit'
-            : `Update to ${
-              bookingDuration === 0.5 ? 'Half' : bookingDuration
-            } ${bookingDuration > 1 ? 'Days' : 'Day'}`,
+          label: isEventCancelled ? 'Cancelled' : rejectionMessages ? 'Submit' : `Update to ${
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/index.js]: Nested ternaries are very hard to read, can you split this out into a function call or something.

File: src/components/BookingModal/BookingModalForm/index.js [hunk 11]
```diff
@@ -44,19 +49,17 @@ const BookingModalForm = props => {
...
+            bookingDuration === 0.5 ? 'Half' : bookingDuration
+          } ${bookingDuration > 1 ? 'Days' : 'Day'}`,
           event: updateEvent,
-          disabled: submitButtonDisabled,
+          disabled: isEventCancelled ? isEventCancelled : submitButtonDisabled,
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/index.js]: `disabled: isEventCancelled || submitButtonDisabled`

File: src/components/BookingModal/BookingModalForm/index.js [hunk 12]
```diff
@@ -76,62 +79,67 @@ const BookingModalForm = props => {
   return (
     <Fragment>
       <Form formData={formData} formStatus={formStatus} actions={createCtas()}>
-        <Input
+        {!rejectionMessages && <Input
           type="select"
           htmlAttrs={{
             name: 'eventTypeId',
             options: renderWFH(),
+
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/index.js]: Delete

File: src/components/BookingModal/BookingModalForm/index.js [hunk 13]
```diff
@@ -76,62 +79,67 @@ const BookingModalForm = props => {
...
             type: 'input',
             name: 'start',
             placeholder: 'Enter a start date',
           }}
           value={formData.start}
+          disabled={isEventCancelled}
           rules={{
             dateNotInPast: true,
           }}
+
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/index.js]: Delete

**Review by bgribz — CHANGES_REQUESTED** [review]

File: src/components/BookingModal/BookingModalForm/index.js (hunk 10 above)
Comment (d1wilko) [inline src/components/BookingModal/BookingModalForm/index.js]: this double ternary operator is super confusing

**Review by d1wilko — CHANGES_REQUESTED** [review]
Alot of good code, one concern for me - which may have been discussed previously

**Comment by bgribz** [comment]
Is there a case here for splitting the content of the form into two or three "FormBody" components rather than conditionally rendering every element that may appear?

**Review by William-Unosquare — APPROVED** [review]
Looks good.

There's a couple of issues remaining with this which we are going to look into today. Namely what Barry had brought up about splitting the form into different parts.

File: src/components/BookingModal/BookingModalForm/index.js [hunk 14]
```diff
@@ -36,6 +43,15 @@ const BookingModalForm = props => {
       : 'Day'
 }`;
 
+    const labelTextValue = `${
+      isEventCancelled
+        ? 'Cancelled'
+        : rejectionMessages
+          ? 'Submit'
+          : `Update to ${bookingDuration === 0.5 ? 'Half' : bookingDuration}
+      ${bookingDuration > 1 ? 'Days' : 'Day'}`
```
Comment (bgribz) [inline src/components/BookingModal/BookingModalForm/index.js]: This still isn't very clear haha, nesting ternaries wherever they may appear isn't very readable.

```
const getLabelTextValue = () => {
  if (isEventCancelled) { return 'Cancelled'; }
  if (rejectionMessages) { return 'Submit'; }

  const durationText = bookingDuration === 0.5 ? 'Half' : bookingDuration;
  const dayText = bookingDuration > 1 ? 'Days' : 'Day';

  return `Update to ${durationText} ${dayText}`;
}
```

**Review by bgribz — CHANGES_REQUESTED** [review]

**Review by bgribz — APPROVED** [review]
